### DISCUSSION TITLE
forecast MARK rules - question about order

### DISCUSSION THREAD

**Discussion by xnr13** [strongswan/strongswan#2448] [General]
Any reason why **iptc_insert_entry** is being used for forecast rule insertion [here](https://github.com/strongswan/strongswan/blob/cf7fb47788dfb83bb5d8bd0bffdb582e381a2f0a/src/libcharon/plugins/forecast/forecast_listener.c#L153) and not **iptc_append_entry**?

My thoughts:
MARK is a non-terminating target of iptables, so rules located lower override those applied previously.
Zombie tunnels caused by instable connection that share everything (IPs, ports) except for the mark value prevent newly-established tunnels from functioning, because they have their marks applied later down the ruleset.

For example, i have marked irrelevant rules with [ZOMBIE RULE] where i am supposed to only have one tunnel active with mark 0x2:
```
-A PREROUTING -d 192.168.111.247/32 -j MARK --set-xmark 0x2/0xffffffff
-A PREROUTING -s 194.44.243.15/32 -d 192.168.1.6/32 -p udp -m udp --sport 1026 --dport 4500 -j MARK --set-xmark 0x2/0xffffffff
-A PREROUTING -d 192.168.111.247/32 -j MARK --set-xmark 0x1/0xffffffff [ZOMBIE RULE]
-A PREROUTING -s 194.44.243.15/32 -d 192.168.1.6/32 -p udp -m udp --sport 1026 --dport 4500 -j MARK --set-xmark 0x1/0xffffffff  [ZOMBIE RULE]
-A OUTPUT -d 192.168.111.247/32 -j MARK --set-xmark 0x2/0xffffffff
-A OUTPUT -d 192.168.111.247/32 -j MARK --set-xmark 0x1/0xffffffff [ZOMBIE RULE]
```

At the moment i have to re-insert all 3 rules with "-A" in an updown script which certainly is a mess

```
iptables-legacy -t mangle -A PREROUTING -d ${PLUTO_PEER_SOURCEIP} -j MARK --set-xmark ${PLUTO_MARK_IN}
iptables-legacy -t mangle -A PREROUTING -s ${PLUTO_PEER} -d ${PLUTO_ME} -p udp -m udp --sport ${PLUTO_UDP_ENC} --dport 4500 -j MARK --set-xmark ${PLUTO_MARK_IN}
iptables-legacy -t mangle -A OUTPUT -d ${PLUTO_PEER_SOURCEIP} -j MARK --set-xmark ${PLUTO_MARK_IN}
```

**Comment by tobiasbrunner** [comment]
> Any reason why **iptc_insert_entry** is being used for forecast rule insertion [here](https://github.com/strongswan/strongswan/blob/cf7fb47788dfb83bb5d8bd0bffdb582e381a2f0a/src/libcharon/plugins/forecast/forecast_listener.c#L153) and not **iptc_append_entry**?

I'd assume to prefer these rules over existing rules in those chains. In particular in `OUTPUT` there might be one or more `REJECT` rule at the end that could prevent the rules from getting applied.

I suppose one possible option could be to use our own chain where we append rules and to which we redirect traffic via an inserted rule in the main chain.

> Zombie tunnels caused by instable connection that share everything (IPs, ports) except for the mark value 

Could you please go into more detail here? What are these zombie tunnels and how are they created?

**Reply by xnr13** [reply]
> > Zombie tunnels caused by instable connection that share everything (IPs, ports) except for the mark value
> 
> Could you please go into more detail here? What are these zombie tunnels and how are they created?

For some reason, probably unstable connections, we often get multiple simultaneously established tunnels for same EAP credentials hanging around on the gateway even though unique=replace is configured in swanctl.conf.
